STRATUM.md

```diff
@@ -0,0 +1,366 @@
+
+### mining.set_target
+
+The target difficulty for a block can change and a server needs to be able to
+notify clients of that.
+
+This method call will only be executed by the server.
+
+
+#### Request
+
+```
+{"id": null, "method": "mining.set_target", "params": ["FFFF000000000000000000000000000000000000000000000000000000000000"]}\n
+```
+
+- [ `id` : `int` ]: request id
+- [ `method` : `string` ]: RPC method name
+- [ `params` : (`string`) ]: list of method parameters
+	1. The the 256bit big-endian target, which MUST be hex encoded.
+
+Any subsequent jobs started by a client after receiving this update MUST
+honor the new target and servers MUST reject submission above this target.
+
+For jobs started before this update, a server MAY accept submissions if they are
+below the previous target.
+
+Please consult the document describing the consensus and proof-of-work
+algorithms for details on the target.
+
+
+#### Response
+
+There is no explicit response for this call.
+
+
+### mining.notify
+
+The notify call is used to supply a worker with new work packages.
+
+This method call will only be executed by the server.
+
+
+#### Request
+
+```
+{"id": null, "method": "mining.notify", "params": ["d70fd222", "abad8f99f3918bf903c6a909d9bbc0fdfa5a2f4b9cb1196175ec825c6610126c", true ]}\n
+```
+
+- [ `id` : `int` ]: request id
+- [ `method` : `string` ]: RPC method name
+- [ `params` : (`string`, `int`, `string`, `bool`) ]: list of method parameters
+	1. Job ID
+	2. Block version
+	3. Header hash
+	4. A boolean indicating whether the miners job queue should be emptied or not
+
+In order to keep the protocol flexible and anticipate changes in both the
+structure of blocks and proof of work scheme, the block version should be
+considered as a switch for the subsequent parameters.
+
+The following is valid for block version (***TODO:*** insert genesis block version).
+
+In its current iteration, all that is needed as input for the proof of work
+puzzle solver is a hash of the block header plus a nonce, as described in the
+consensus document.
+Thus a client only requires the header hash from the server.
+
+
+#### Response
+
+There is no explicit response for this call.
+
+
+### mining.submit
+
+With this method a worker can submit solutions for the mining puzzle.
+This method call will only be executed by clients.
+
+
+#### Request
+
+```
+{"id": 4, "method": "mining.submit", "params": ["WORKER_NAME", "d70fd222", "98b6ac44d2", ""]}\n
+```
+
+- [ `id` : `int` ]: request id
+- [ `method` : `string` ]: RPC method name
+- [ `params` : (`string`, `string`, `string`, `solution`) ]: list of method
+  parameters
+	1. Worker name, mostly for statistics
+	2. Job ID
+	3. Miner nonce in hex format
+	4. Proof of work puzzle solution
+
+As with `mining.notify` all parameters after the job ID should be understood as
+being conditional on the block version sent along in `mining.notify`.
+
+The miner nonce MUST be `8 - len(extra_nonce)` bytes long and be little endian.
```

Comment:
I got confused here, the extranonce was set by the server, why have the length of this field being based on the extra_nonce?

Reply:
The miner and extra nonce are concatenated to the full nonce. So if the server sets the extra nonce to be 3 bytes long then the miner must submit 5 bytes nonce with this submit call.

Reply:
Is it ok to have just 8 bytes for nonce? For FPGA for example or for potential ASICs? Usually pools will lease at least 2 bytes (likely 4 bytes) for extra nonce. So only 4 bytes left for miner which could be not enough if miner is very fast (10 GPU set for example).

Reply:
My proposal is to set it to 32 bytes while it's not too late.

Reply:
The nonces in our block headers are currently 8 bytes, since there's really no need for a bigger nonce with cuckoo cycle—at least given our current knowledge of available hardware—,although the algorithm has space for the nonces to be bigger.

If we consider a bad case of not producing a block for double the normal block time, i.e. 6 minutes, and a solution taking about 100ms to find, you'd be able to generate about 3600 solutions with a single GPU. At 3600 solutions you'd need to run about 1.2 million GPUs to exhaust those 4bytes.

But, there's a case to be made that future PoW changes might require us to make the nonce space bigger.

Reply:
My point is that 4 bytes will be taken by pool leaving only 4 bytes for miner. Why not just extend it once and forever, extra few bytes in a block is not a problem, a HF to meet hw requirements is. If AE will be a successful project we will see ASIC in 1-2 years, current ASIC performance over CPU on equihash for example is dramatic, so it probably wise to just use 32 bytes since the start and don't bother any more, no need for pools to upgrade consensus in the future, no need for HF, etc.

Reply:
Is there any epoch RPC endpoints responsible for block templates?